stan/math/opencl/kernel_generator/reduction_2d.hpp

```diff
@@ -0,0 +1,342 @@
+template <typename Derived, typename T, typename Operation>
+class reduction_2d
+    : public internal::reduction_2d_base,
+      public operation_cl<Derived, typename std::remove_reference_t<T>::Scalar,
+                          T> {
+ public:
+  using Scalar = typename std::remove_reference_t<T>::Scalar;
+  using base = operation_cl<Derived, Scalar, T>;
+  using base::var_name_;
+  static const bool require_specific_local_size = true;
+
+ protected:
+  std::string init_;
+  using base::derived;
+
+ public:
+  using base::cols;
+  /**
+   * Constructor
+   * @param a the expression to reduce
+   * @param init OpenCL source code of initialization value for reduction
+   */
+  explicit reduction_2d(T&& a, const std::string& init)
+      : base(std::forward<T>(a)), init_(init) {}
+
+  /**
+   * Generates kernel code for assigning this expression into result expression.
+   * @param[in,out] generated map from (pointer to) already generated local
+   * operations to variable names
+   * @param[in,out] generated_all map from (pointer to) already generated all
+   * operations to variable names
+   * @param ng name generator for this kernel
+   * @param row_index_name row index variable name
+   * @param col_index_name column index variable name
+   * @param result expression into which result is to be assigned
+   * @return part of kernel with code for this and nested expressions
+   */
+  template <typename T_result>
+  kernel_parts get_whole_kernel_parts(
+      std::map<const void*, const char*>& generated,
+      std::map<const void*, const char*>& generated_all, name_generator& ng,
+      const std::string& row_index_name, const std::string& col_index_name,
+      const T_result& result) const {
+    kernel_parts parts = derived().get_kernel_parts(
+        generated, generated_all, ng, row_index_name, col_index_name, false);
+    kernel_parts out_parts = result.get_kernel_parts_lhs(
+        generated, generated_all, ng, row_index_name, col_index_name);
+
+    parts.args += out_parts.args;
+    parts.reduction_2d += "if (lid_i == 0) {\n"
+                     + result.var_name_
+                     + "_global[wg_id_j * n_groups_i + wg_id_i] = "
+                     + derived().var_name_ + "_local[0];\n"
+                     "}\n";
+    return parts;
+  }
+
+  /**
+   * Generates kernel code for this and nested expressions.
+   * @param row_index_name row index variable name
+   * @param col_index_name column index variable name
+   * @param view_handled whether whether caller already handled matrix view
+   * @param var_name_arg name of the variable in kernel that holds argument to
+   * this expression
+   * @return part of kernel with code for this and nested expressions
+   */
+  inline kernel_parts generate(const std::string& row_index_name,
+                               const std::string& col_index_name,
+                               const bool view_handled,
+                               const std::string& var_name_arg) const {
```

Comment:
[optional] If these are very small strings then totally no need, but if they are larger than short string optimization can handle should these be perfect forwarding references to move them?

Reply:
No, perfect forwarding does not help here. Comparing to const ref, moving is just one more operation we do not need here.